solarwinds_apm/otlp_metrics_processor.py

```diff
@@ -0,0 +1,190 @@
+# © 2023 SolarWinds Worldwide, LLC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at:http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+import logging
+import random
+from typing import TYPE_CHECKING, Any, Tuple
+
+from opentelemetry.metrics import get_meter_provider
+from opentelemetry.sdk.trace import SpanProcessor
+from opentelemetry.semconv.trace import SpanAttributes
+from opentelemetry.trace import SpanKind, StatusCode
+
+from solarwinds_apm.w3c_transformer import W3CTransformer
+
+if TYPE_CHECKING:
+    from opentelemetry.sdk.trace import ReadableSpan
+
+    from solarwinds_apm.apm_config import SolarWindsApmConfig
+    from solarwinds_apm.apm_meter_manager import SolarWindsMeterManager
+    from solarwinds_apm.apm_txname_manager import SolarWindsTxnNameManager
+
+
+logger = logging.getLogger(__name__)
+
+
+class SolarWindsOTLPMetricsSpanProcessor(SpanProcessor):
+    # TODO Refactor for both inbound and otlp metrics
+    #      https://swicloud.atlassian.net/browse/NH-65061
+    _HTTP_METHOD = SpanAttributes.HTTP_METHOD  # "http.method"
+    _HTTP_ROUTE = SpanAttributes.HTTP_ROUTE  # "http.route"
+    _HTTP_STATUS_CODE = SpanAttributes.HTTP_STATUS_CODE  # "http.status_code"
+    _HTTP_URL = SpanAttributes.HTTP_URL  # "http.url"
+
+    _HTTP_SPAN_STATUS_UNAVAILABLE = 0
+
+    def __init__(
+        self,
+        apm_txname_manager: "SolarWindsTxnNameManager",
+        apm_config: "SolarWindsApmConfig",
+        apm_meters: "SolarWindsMeterManager",
+    ) -> None:
+        self.apm_txname_manager = apm_txname_manager
+        self.service_name = apm_config.service_name
+        self.apm_meters = apm_meters
+
+    # TODO Assumes SolarWindsInboundMetricsSpanProcessor.on_start
+    #      is called to store span ID for any custom txn naming
+    #      https://swicloud.atlassian.net/browse/NH-65061
+
+    def on_end(self, span: "ReadableSpan") -> None:
+        """Calculates and reports OTLP trace metrics"""
+        # Only calculate OTLP metrics for service entry spans
+        parent_span_context = span.parent
+        if (
+            parent_span_context
+            and parent_span_context.is_valid
+            and not parent_span_context.is_remote
+        ):
+            return
+
+        # support ssa and conform to Otel proto common_pb2
+        meter_attrs = {
+            "sw.service_name": self.service_name,
```

Comment:
this is to conform to the [SWO response time metric tagging](https://github.com/librato/trace/blob/master/docs/specs/agent-metrics.md#swo-tags), right? If so, `sw.service_name` isn't needed because Python doesn't support service name override.

Reply:
Ahhhh. Removed in [f432626](https://github.com/solarwindscloud/solarwinds-apm-python/pull/201/commits/f43262622aae77fd1340737ddad2bd9c63ebe39b). Past Tammy also might have added it when trying to correlate metrics with service, before you found that it was that resource attribute needed instead!